### ISSUE TITLE
SoftmaxBackward NaN issue

### ISSUE THREAD

**Issue by VarunGumma** [facebookresearch/fairseq#4936]
#### What have you tried?
I have been trying to train large Transformer Models (with 200-400M parameters) with `GELU`, `layernorm-embedding` and `normalize-before` and the training constantly crashes when I use `fp16`. It runs quite well with `fp32`, but the training is dead slow. Any workarounds for this?

#### What's your environment?

 - fairseq Version (e.g., 1.0 or main): `0.12.2`
 - PyTorch Version (e.g., 1.0): `1.13.0`
 - OS (e.g., Linux):  `NVIDIA DGX Server Version 5.0.5`
 - How you installed fairseq (`pip`, source): `conda`
 - Build command you used (if compiling from source): -
 - Python version: `3.10.6`
 - CUDA/cuDNN version: `Cuda11.3`
 - GPU models and configuration: `A100`
 - Any other relevant information: -

**Comment by VarunGumma** [comment]
Found a fix:

Following [this](https://github.com/kenziyuliu/MS-G3D/issues/40), I removed anomaly detection (set `torch.autograd.set_detect_anomaly(False)` in `fairseq_cli/train.py`) and found a 3x speedup and better `fp16` handling (no crashes, but the loss scale oscillates quite a bit because of this, which is not an issue).
